src/System.Text.Primitives/System/Text/Encoders/Utf8Validator.cs

```diff
@@ -0,0 +1,252 @@
+                return ConsumeDataWithoutExistingPartialSequence(utf8Bytes.Slice(numBytesRequiredFromInputBufferToFinishPartialSequence));
+            }
+            else if (validity == SequenceValidity.Incomplete)
+            {
+                // We've consumed all data available to us and we still have an incomplete sequence.
+                // It's still valid (until we see invalid bytes), so squirrel away what we've seen
+                // and report success to our caller.
+
+                Debug.Assert(numBytesConsumed < 4);
+                Debug.Assert(numBytesConsumed == partialSequenceNewByteCount);
+
+                // Put all partial data into the high 3 bytes, making room for us to
+                // write the count of partial bytes in the buffer as the low byte.
+
+                if (BitConverter.IsLittleEndian)
+                {
+                    return (partialSequence << 8) | (uint)numBytesConsumed;
+                }
+                else
+                {
+                    return (partialSequence & unchecked((uint)~0xFF)) | (uint)numBytesConsumed;
+                }
+            }
+            else
+            {
+                // Truly invalid data.
+                // (Shouldn't have gotten 'Empty' or 'WellFormed'.)
+
+                Debug.Assert(validity == SequenceValidity.Invalid);
+
+                return InvalidSequence;
+            }
+        }
+
+        private static unsafe uint ConsumeDataWithoutExistingPartialSequence(ReadOnlySpan<byte> utf8Bytes)
+        {
+            var indexOfFirstInvalidSequence = Utf8Utility.GetIndexOfFirstInvalidUtf8Sequence(utf8Bytes);
+            if (indexOfFirstInvalidSequence < 0)
+            {
+                // Successfully consumed entire buffer without error.
+
+                return DefaultSequence;
+            }
+            else
+            {
+                // Couldn't consume entire buffer; is this due to a partial buffer or truly invalid data?
+
+                utf8Bytes = utf8Bytes.Slice(indexOfFirstInvalidSequence);
+                var validity = Utf8Utility.PeekFirstSequence(utf8Bytes, out int numBytesConsumed, out _);
+                if (validity == SequenceValidity.Incomplete)
+                {
+                    // Saw a partial (not invalid) sequence, remember it for next time.
+
+                    Debug.Assert(1 <= numBytesConsumed && numBytesConsumed <= 3);
+
+                    // Put all partial data into the high 3 bytes, making room for us to
+                    // write the count of partial bytes in the buffer as the low byte.
+
+                    if (BitConverter.IsLittleEndian)
+                    {
+                        uint partialSequence = (uint)numBytesConsumed;
+                        utf8Bytes.Slice(0, numBytesConsumed).CopyTo(new Span<byte>(&partialSequence, sizeof(uint)).Slice(1));
```

Comment:
Same here. This method could be safe if you use the binary APIs.

Reply:
Fixed.